spec/prog/vm/nexus_spec.rb

```diff
@@ -5,42 +5,117 @@
+
+    before do
+      expect(vm).to receive(:vm_host).and_return(vm_host)
+    end
+
+    it "runs adduser" do
+      expect(sshable).to receive(:cmd).with(/sudo.*adduser.*#{nx.vm_name}/)
+
+      expect { nx.create_unix_user }.to raise_error Prog::Base::Hop do |hop|
+        expect(hop.new_label).to eq("prep")
+      end
+    end
+
+    it "absorbs an already-exists error as a success" do
+      expect(sshable).to receive(:cmd).with(/sudo.*adduser.*#{nx.vm_name}/).and_raise(
+        Sshable::SshError.new("adduser: The user `vmabc123' already exists.")
+      )
+
+      expect { nx.create_unix_user }.to raise_error Prog::Base::Hop do |hop|
+        expect(hop.new_label).to eq("prep")
+      end
+    end
+
+    it "raises other errors" do
+      ex = Sshable::SshError.new("out of memory")
+      expect(sshable).to receive(:cmd).with(/sudo.*adduser.*#{nx.vm_name}/).and_raise(ex)
+
+      expect { nx.create_unix_user }.to raise_error ex
+    end
+  end
+
+  describe "#prep" do
+    it "generates and passes a params json" do
+      vm = nx.vm
+      vm.ephemeral_net6 = "fe80::/64"
+      vm.unix_user = "test_user"
+      vm.public_key = "test_ssh_key"
+      expect(vm).to receive(:private_subnets).and_return [NetAddr.parse_net("fd10:9b0b:6b4b:8fbb::/64")]
+      expect(vm).to receive(:cloud_hypervisor_cpu_topology).and_return(Vm::CloudHypervisorCpuTopo.new(1, 1, 1, 1))
+
+      sshable = instance_spy(Sshable)
+      vmh = instance_double(VmHost, sshable: sshable,
+        total_cpus: 80, total_cores: 80, total_nodes: 1, total_sockets: 1, ndp_needed: false)
+      expect(vm).to receive(:vm_host).and_return(vmh)
+
+      expect(sshable).to receive(:cmd).with(/echo (.|\n)* \| sudo -u vm[0-9a-z]+ tee/) do
+        require "json"
+        params = JSON(_1.shellsplit[1])
+        expect(params).to include({
+          "public_ipv6" => "fe80::/64",
+          "unix_user" => "test_user",
+          "ssh_public_key" => "test_ssh_key",
+          "max_vcpus" => 1,
+          "cpu_topology" => "1:1:1:1",
+          "mem_gib" => 4
+        })
+      end
+      expect(sshable).to receive(:cmd).with(/sudo bin\/prepvm/)
+
+      expect { nx.prep }.to raise_error Prog::Base::Hop do |hop|
+        expect(hop.new_label).to eq("trigger_refresh_mesh")
+      end
+    end
+  end
+
+  describe "#start" do
+    it "allocates the vm to a host" do
+      vmh = VmHost.new(
+        net6: NetAddr.parse_net("2a01:4f9:2b:35a::/64"),
+        ip6: NetAddr.parse_ip("2a01:4f9:2b:35a::2")
+      ) { _1.id = "46ca6ded-b056-4723-bd91-612959f52f6f" }
+
+      expect(nx).to receive(:allocate).and_return("46ca6ded-b056-4723-bd91-612959f52f6f")
+      expect(VmHost).to receive(:[]).with("46ca6ded-b056-4723-bd91-612959f52f6f") { vmh }
+      expect(vm).to receive(:update) do |**args|
+        expect(args[:ephemeral_net6]).to match(/2a01:4f9:2b:35a:.*/)
+        expect(args[:vm_host_id]).to match "46ca6ded-b056-4723-bd91-612959f52f6f"
+      end
```

Comment:
nit: I feel like too many hardcoded ID decreases readability. They are long and my brain tries to compare them while reading.
```suggestion
      vmh_id = "46ca6ded-b056-4723-bd91-612959f52f6f"
      vmh = VmHost.new(
        net6: NetAddr.parse_net("2a01:4f9:2b:35a::/64"),
        ip6: NetAddr.parse_ip("2a01:4f9:2b:35a::2")
      ) { _1.id = vmh_id }

      expect(nx).to receive(:allocate).and_return(vmh_id)
      expect(VmHost).to receive(:[]).with(vmh_id) { vmh }
      expect(vm).to receive(:update) do |**args|
        expect(args[:ephemeral_net6]).to match(/2a01:4f9:2b:35a:.*/)
        expect(args[:vm_host_id]).to match vmh_id
      end
```

Reply:
Sure, we can use that style. I used to use SecureRandom, but @will found that it was meaningful in test performance at some later date on another project.